front: add reset a single occurrence action

SharglutDev opened the pull request:
close #11469

SharglutDev committed:
front: add reset single occurrence action

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -147,10 +149,68 @@ const useOccurrenceActions = ({
+        throw new Error('Cannot reset an occurrence which was not an exception');
+      }
+
+      let updatedExceptions = [...pacedTrain.exceptions];
```
nit: do not initialize updatedExceptions right now and use `pacedTrain.exceptions` below
```suggestion
      let updatedExceptions: PacedTrainException[];
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -147,10 +149,68 @@ const useOccurrenceActions = ({
+          updatedExceptions,
+          // If it is an indexed occurrence, the corresponding exception will always have an occurrence_index
+          occurrenceToUpdateException.occurrence_index!
+        );
```
You are using the index in the occurrences array to map in the exceptions array, so it can't work :)
It would be simpler to do a filter here

SharglutDev replied:
It weirdly worked because we always add added exceptions at the end of the array but it was very flaky good catch :) Fixed.

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -147,10 +149,68 @@ const useOccurrenceActions = ({
+   */
+  const resetOccurrenceExceptions = useCallback(
+    (occurrenceId: OccurrenceId) => {
+      const occurrenceToUpdateException = findExceptionWithOccurrenceId(
```
nit
```suggestion
      const exceptionToUpdate = findExceptionWithOccurrenceId(
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -147,10 +149,68 @@ const useOccurrenceActions = ({
+        );
+      } else {
+        // update occurrenceToUpdateException by removing all its properties except key and start time
+        const resetAddedException: PacedTrainException = {
+          key: occurrenceToUpdateException.key,
+          start_time: occurrenceToUpdateException.start_time,
+        };
+        const exceptionIndex = updatedExceptions.findIndex(
+          ({ key }) => resetAddedException.key === key
+        );
+        updatedExceptions = replaceElementAtIndex(
+          updatedExceptions,
+          exceptionIndex,
+          resetAddedException
+        );
+      }
```
You can also simply use a map
```suggestion
        updatedExceptions = pacedTrain.exceptions.map((exception) =>
          exception.key === exceptionToUpdate.key
            ? {
                key: exceptionToUpdate.key,
                start_time: exceptionToUpdate.start_time,
              }
            : exception
        );
```

SharglutDev replied:
Much better :) Done

SharglutDev committed:
front: fix restore action button for added exceptions

The button should be displayed only if there is more than start_time change group in the exception

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -147,10 +148,60 @@ const useOccurrenceActions = ({
+      storePacedTrain(
+        pacedTrain.id,
+        formattedPacedTrain,
+        timetableId!,
```
It can be done in another PR, but you can get the `timetableId` from the `ScenarioContext` and then remove the !

SharglutDev replied:
Oh right, I'll do that for all actions in the file in another PR :)

clarani reviewed (approved):
LGTM & tested :white_check_mark:
